pkg/services/ngalert/notifier/autogen_alertmanager.go

```diff
@@ -0,0 +1,206 @@
+	if err != nil {
+		return err
+	}
+
+	err = autogenRoute.addToRoute(cfg.GetRoute())
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// newAutogeneratedRoute creates a new autogenerated route based on the notification settings for the given org.
+// cfg is used to construct the settings validator and to ensure we create a dedicated route for each receiver.
+// skipInvalid is used to skip invalid settings instead of returning an error.
+func newAutogeneratedRoute(ctx context.Context, logger log.Logger, store autogenRuleStore, orgId int64, cfg apiAlertingConfig, skipInvalid bool) (autogeneratedRoute, error) {
+	settings, err := store.ListNotificationSettings(ctx, orgId)
+	if err != nil {
+		return autogeneratedRoute{}, fmt.Errorf("failed to list alert rules: %w", err)
+	}
+
+	notificationSettings := make(map[data.Fingerprint]models.NotificationSettings)
+	// Add a default notification setting for each contact point. This is to ensure that we always have a route for each
+	// contact point even if no rules are using it. This will prevent race conditions between AM sync and rule sync.
+	for receiver := range cfg.ReceiverNames() {
+		setting := models.NewDefaultNotificationSettings(receiver)
+		fp := setting.Fingerprint()
+		notificationSettings[fp] = setting
+	}
+
+	validator := NewNotificationSettingsValidator(cfg)
+	for ruleKey, ruleSettings := range settings {
+		for _, setting := range ruleSettings {
+			// TODO we should register this errors and somehow present to the users or make sure the config is always valid.
+			if err = validator.Validate(setting); err != nil {
+				if skipInvalid {
+					logger.Error("Rule notification settings are invalid. Skipping", append(ruleKey.LogContext(), "error", err)...)
+					continue
+				}
+				return autogeneratedRoute{}, fmt.Errorf("invalid notification settings for rule %s: %w", ruleKey.UID, err)
+			}
+			fp := setting.Fingerprint()
+			// Keep only unique settings.
+			if _, ok := notificationSettings[fp]; ok {
+				continue
+			}
+			notificationSettings[fp] = setting
+		}
+	}
+	if len(notificationSettings) == 0 {
+		return autogeneratedRoute{}, nil
+	}
+	newAutogenRoute, err := generateRouteFromSettings(cfg.GetRoute().Receiver, notificationSettings)
+	if err != nil {
+		return autogeneratedRoute{}, fmt.Errorf("failed to create autogenerated route: %w", err)
+	}
+	return newAutogenRoute, nil
+}
+
+type autogeneratedRoute struct {
+	Route       *definitions.Route
+	Fingerprint data.Fingerprint
+}
+
+// generateRouteFromSettings generates a route and fingerprint for this route. The route is a tree of 3 layers:
+//  1. with matcher by label models.AutogeneratedRouteLabel equals 'true'.
+//  2. with matcher by receiver name.
+//  3. with matcher by unique combination of optional settings. It is created only if there are optional settings.
+func generateRouteFromSettings(defaultReceiver string, settings map[data.Fingerprint]models.NotificationSettings) (autogeneratedRoute, error) {
```

Comment:
It really feels like this should be bigger for all it does, but that's it. Had to read through a couple times because it felt like we had to be recursing somewhere 😅